File: backend/app/services/notifier.py
"""Email alert notifier.

Sends HTML+plain-text emails via SMTP (async) with retry logic.
Logs every attempt to notification_logs.
"""

import asyncio
import uuid
from datetime import datetime, timezone
from email.mime.multipart import MIMEMultipart
from email.mime.text import MIMEText

import aiosmtplib
from sqlalchemy.ext.asyncio import AsyncSession

from app.config import settings
from app.logging_config import get_logger
from app.models.alert import Alert
from app.models.notification_log import NotificationLog
from app.models.recipient import Recipient
from app.services.email_renderer import render_alert_email

logger = get_logger(__name__)

_MAX_RETRIES = 3
_RETRY_BACKOFF = [5.0, 10.0, 20.0]


class AlertNotifier:
    async def notify(
        self,
        db: AsyncSession,
        alert: Alert,
        recipients: list[Recipient],
        location_name: str | None,
    ) -> None:
        post = alert.post
        subject, html_body, plain_body = render_alert_email(
            keyword=alert.matched_keyword,
            content=post.content,
            author_name=post.author_name,
            neighborhood=post.neighborhood,
            posted_at=post.posted_at,
            source_url=post.source_url,
            location_name=location_name,
        )

        for recipient in recipients:
            log = NotificationLog(
                id=uuid.uuid4(),
                alert_id=alert.id,
                recipient_id=recipient.id,
                status="pending",
                attempt_count=0,
            )
            db.add(log)

            success = await self._send_with_retry(recipient.email, subject, html_body, plain_body, log)
            if success:
                log.status = "sent"
                log.sent_at = datetime.now(timezone.utc)
                logger.info("alert_sent", recipient=recipient.email, keyword=alert.matched_keyword)
            else:
                log.status = "failed"
                logger.error("alert_failed", recipient=recipient.email, keyword=alert.matched_keyword)

        await db.commit()

    async def _send_with_retry(
        self,
        to_email: str,
        subject: str,
        html_body: str,
        plain_body: str,
        log: NotificationLog,
    ) -> bool:
        msg = MIMEMultipart("alternative")
        msg["Subject"] = subject
        msg["From"] = settings.smtp_from
        msg["To"] = to_email
        msg.attach(MIMEText(plain_body, "plain"))
        msg.attach(MIMEText(html_body, "html"))

        for attempt, backoff in enumerate((*_RETRY_BACKOFF, None), start=1):
            log.attempt_count = attempt
            try:
                await aiosmtplib.send(
                    msg,
                    hostname=settings.smtp_host,
                    port=settings.smtp_port,
                    username=settings.smtp_user or None,
                    password=settings.smtp_password or None,
                    use_tls=False,
                    start_tls=settings.smtp_use_tls,
                )
                return True
            except Exception as exc:
                log.error_message = str(exc)
                logger.warning("smtp_send_failed", attempt=attempt, error=str(exc))
                if backoff is not None:
                    await asyncio.sleep(backoff)

        return False

File: backend/src/services/notifier.js
'use strict';

const nodemailer = require('nodemailer');
const { settings } = require('../config');
const { logger } = require('../logger');
const { renderAlertEmail } = require('./emailRenderer');

const MAX_RETRIES = 3;
const RETRY_BACKOFF_MS = [5000, 10000, 20000];

function sleep(ms) {
  return new Promise((resolve) => setTimeout(resolve, ms));
}

function createTransport() {
  return nodemailer.createTransport({
    host: settings.smtpHost,
    port: settings.smtpPort,
    secure: false,
    requireTLS: settings.smtpUseTls,
    auth: settings.smtpUser
      ? { user: settings.smtpUser, pass: settings.smtpPassword }
      : undefined,
  });
}

/**
 * Sends alert emails to all recipients.
 * db is optional — when provided it is used to persist NotificationLog records.
 */
async function notifyAlert({ alert, recipients, locationName, db }) {
  const post = alert.post;
  const { subject, html, plain } = renderAlertEmail({
    keyword: alert.matched_keyword,
    content: post.content,
    author_name: post.author_name,
    neighborhood: post.neighborhood,
    posted_at: post.posted_at,
    source_url: post.source_url,
    location_name: locationName,
  });

  const transport = createTransport();

  for (const recipient of recipients) {
    let lastError = null;
    let sent = false;

    for (let attempt = 1; attempt <= MAX_RETRIES; attempt++) {
      try {
        await transport.sendMail({
          from: settings.smtpFrom,
          to: recipient.email,
          subject,
          text: plain,
          html,
        });
        sent = true;
        logger.info('alert_sent', { recipient: recipient.email, keyword: alert.matched_keyword });
        break;
      } catch (err) {
        lastError = err.message;
        logger.warning('smtp_send_failed', { attempt, error: err.message });
        if (attempt < MAX_RETRIES) await sleep(RETRY_BACKOFF_MS[attempt - 1]);
      }
    }

    if (db) {
      const { v4: uuidv4 } = require('uuid');
      await db.query(
        `INSERT INTO notification_logs (id, alert_id, recipient_id, status, sent_at, error_message, attempt_count)
         VALUES ($1, $2, $3, $4, $5, $6, $7)`,
        [
          uuidv4(),
          alert.id,
          recipient.id,
          sent ? 'sent' : 'failed',
          sent ? new Date().toISOString() : null,
          sent ? null : lastError,
          MAX_RETRIES,
        ]
      );
    }

    if (!sent) {
      logger.error('alert_failed', { recipient: recipient.email, keyword: alert.matched_keyword, error: lastError });
    }
  }
}

module.exports = { notifyAlert };
